src/geometry_aux.cpp

```diff
@@ -40,6 +42,71 @@ void update_universe_cell_count(int32_t a, int32_t b)
   }
 }
 
+double calculate_optical_thickness(Position start_sampled_position,
+  Position end_sampled_position, uint64_t* seed)
+{
+  Position direction = (end_sampled_position - start_sampled_position) /
+                       (end_sampled_position - start_sampled_position).norm();
+
+  Particle p;
+  SourceSite site;
+  site.E = 1.0;
+  site.particle = ParticleType::neutron;
+  site.r = start_sampled_position;
+  site.u = direction;
+  p.from_source(&site);
+
+  double optical_thickness = 0.0;
+
+  while (true) {
+    if (!exhaustive_find_cell(p))
+      break;
+
+    BoundaryInfo boundary = distance_to_boundary(p);
+    double distance_to_boundary = boundary.distance;
+    double distance_to_endpoint = (end_sampled_position - p.r()).norm();
+
+    // Check if we are in the cell containing the endpoint
+    if (distance_to_endpoint < distance_to_boundary) {
+      // Use the remaining distance to the endpoint
+      p.event_calculate_xs();
+      optical_thickness += distance_to_endpoint * p.macro_xs().total;
+      break; // Stop the loop
+    }
+
+    p.event_calculate_xs();
+    optical_thickness += distance_to_boundary * p.macro_xs().total;
+
+    p.move_distance(distance_to_boundary);
+  }
+
+  return optical_thickness;
+}
+
+extern "C" void openmc_get_optical_thickness(Position start_voxel_min,
```

Comment:
This function and the previous one have similar names now.  Maybe this one can be renamed to better convey what it's doing.  `get_mean_optical_thickness_between_voxels` or something like that?